design/ozone.md

```diff
@@ -0,0 +1,246 @@
+   return err
+}
+
+assert.Contains(t, s, "prometheus.io/scrape: \"true\"")
+```
+
+### Create k8s resource from operator
+
+One easy way to integrate the exising resource generation of Ozone is to use it from golang API. This is similar to the approach of [ceph csi](https://github.com/rook/rook/tree/master/pkg/operator/ceph/csi) which reads the k8s template files from the disk.
+
+The main diference here is that before the final adjustment we can reuse the basic transformations which are already implemented by ozone.
+
+At the end of the transformations we will have a list of real K8s resources object, where we can do any final adjustment on the API object and send them to the k8s API services.
+
+The main logic is the following: 
+
+ * The resource generator can be initialized with the main k8s definition directory (which is included in the Ozone release, and included in the rook operator image)
+ * The transformations can be added programatically based on CRD properties
+ * The real k8s golang objects are generated
+ * After final adjustment in golang, we can post all the objects to the K8S api
+
+### CRD
+
+A good operator CRD provides a good balance between the simplicity and power. By default the CRD should be as simple as possible but we should keep the possibility to customize any segment of the generated cluster.
+
+This can be achieved by the use k8s generator tool.
+
+ 1. The default cluster will be generated based on the main attributes of CRD
+ 2. Additional attributes can trigger additional transformations
+ 3. Finally we can provide an option to define any custom transformations
+
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: rook-ozone
+---
+apiVersion: ozone.rook.io/v1alpha1
+kind: ObjectStore
+metadata:
+  name: my-store
+  namespace: rook-ozone
+spec:
+  components: #high lebel options to turn off/on component groups
+     csi: true
+     recon: true
+     s3g: true
+  features: # these are not components but confiuration groups
+     prometheus: true
+     tracing: true
+  ozoneVersion: #see ceph versioning doc for mor details
+    image: apache/ozone:0.4.1
+    allowUnsupported: false
+  upgradePolicy: #see ceph versioning doc for mor details (Won't be implmented in Phase 1)
+    ozoneVersion:
+      image: apache/ozone:0.5.0
+      allowUnsupported: false
+  storage: #more storage option should be added to here
+     type: ephemeral/auto/manual # ephemeral: use emptyDir, auto: using the existing rook method to identify local disk, manual: should be specified
+     persistence:
+        metadata: #specify persistent volumes for metadata disk
+        storage: #specify persistent volumes for storage
+  annotations: #custom annotations for all the created k8s resources
+     key: value
+  customization: #deep water: any additional transformations to modify k8s resources. Only for advanced users.
+    - type: Add
+      path: ["metadata", "labels"]
+      value:
+         generated-with: rook
+    
```

Comment:
Also consider adding a `placement` setting to allow specifying the node selector, affinity, and anti-affinity. Most commonly for other providers such as ceph, it is used to allow specifying on which nodes the daemons should be started.

Reply:
Sure. I added placement to here and to the component levels (eg. component/scm/placement).